### ISSUE TITLE
The tfrecords file is 8 times larger than raw image data

### ISSUE THREAD

**Issue by daiab** [tensorflow/tensorflow#9675]
I try to write a tfrecords file, but the file is larger than raw data. 
```python (type)
img = Image.open('img_file')  # this image file size: 24 kb
b = img.tobytes()  # the len(b) is 24 kb, this is right
feats = tf.train.Features(feature={'label': tf.train.Feature(int64_list=tf.train.Int64List(value=[2])), 
'image_raw': tf.train.Feature(bytes_list=tf.train.BytesList(value=[b]))})
example_string = example.SerializeToString()
len(example_string) / 8 / 1024  # the output == 24.0057373046875 kb, look like well
```
but I write this 'example_string' to tfrecords file , the tfrecords file size become 192 kb, I cann`t understand why tfrecords file size serval times larger than 'example_string' and raw image data

**Comment by ppwwyyxx** [comment]
Because you use Int64List.

**Comment by asanakoy** [comment]
@ppwwyyxx , but he uses BytesList for the image, not  Int64. 
The only int64 value in 1 integer label.
I have the problem of blowing up tfrecords as well. My images are 30G when wtored in jpg, but 280G when I write them in tfrecord 

        example = tf.train.Example(features=tf.train.Features(feature={
            'height': _int64_feature(height), # single integer
            'width': _int64_feature(width),  # single integer
            'image_raw': _bytes_feature(img_raw), # image
            'image_id': _int64_feature(image_id)}))  # single integer

**Comment by ppwwyyxx** [comment]
`30G when wtored in jpg, but 280G when I write them in tfrecord`.. do you mean they are __not__ jpeg when stored in tfrecord?

**Comment by asanakoy** [comment]
@ppwwyyxx , Probably the are not jpeg anymore when stored in tfrecord. 
When I used leveldb binaries with caffe, it didn't blow up so much.
Looks like it stores raw, without any encoding?

**Comment by ppwwyyxx** [comment]
Then it definitely will be several times larger and the factor depends on how well JPEG works on your images. The factor is 5.x on ImageNet, btw.

**Comment by asanakoy** [comment]
Is it because tfrecords stores decompressed images? 
Is it possible to to store compressed? Do you think it will make everything much slower?

**Comment by ppwwyyxx** [comment]
TFRecord stores bytes so you can do any encoding you want.
The easiest thing you can do is just `img_raw = open(img_path).read()` to use the original encoding.
The decompression time is usually much smaller than training, and could be completely hidden if the preprocessing runs in parallel with training.

**Comment by laura-wang** [comment]
@ppwwyyxx  This solution really reduces the size. But  when use` image = tf.decode_raw(features['image_raw'], tf.uint8)` to decode the image, the image size is much smaller. Do you know how can I read from the tfrecord to get the original image?

**Comment by laura-wang** [comment]
I find the solution
`tf.image.decode_jpeg(image_raw_data)` or
`cv2.imdecode(np.fromstring(image_raw_data, dtype=np.uint8), -1)`
will solve the problem.

**Comment by switchfootsid** [comment]
For anyone who is confused how serialization of digital images works, this is a pretty wonderful explanation of TFRecords https://planspace.org/20170403-images_and_tfrecords/

**Comment by FalsoMoralista** [comment]
Is it really necessary to store image dimensions as features for handling datasets with variable dimension images?
